In [ ]:
# prompt: load libraries for data exploration and visualization

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import altair as alt

In [ ]:
# prompt: read all the csv in /content/sample_data/app_file to different df

import pandas as pd
import os

directory_path = '/content/sample_data/app_file'
dfs = {}

# List all files in the directory
all_files = os.listdir(directory_path)

# Filter for CSV files and store them in a list
csv_files = [f for f in all_files if f.endswith('.csv')]

# Read each CSV file into a separate DataFrame and store it in the dictionary
for csv_file in csv_files:
    file_path = os.path.join(directory_path, csv_file)
    # Create a descriptive key for the dictionary (e.g., filename without extension)
    df_name = os.path.splitext(csv_file)[0]
    dfs[df_name] = pd.read_csv(file_path)

# You can now access each DataFrame by its name, e.g., dfs['california_housing_train']
print(dfs.keys()) # To see the names of the dataframes
# To see the head of a specific dataframe


dict_keys(['Response_survey_rows (2)', 'Response_rows (2)'])


In [ ]:
dfs['Response_rows (2)'].head()

id                     created_at                               user_id  \
0  336  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
1  341  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
2  343  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
3  344  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
4  342  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   

                    response_project_id  response_policy_id response_group  \
0  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   1     allocation   
1  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   6     allocation   
2  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   8     allocation   
3  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   9     allocation   
4  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   7     allocation   

   response_value  response_bts_value  
0               5                  70  
1               2                  20  
2               3                  10  
3               0                   0  
4               0                   0

In [ ]:
# prompt: split each of the df in dfs into different df based on (response_group) columns

grouped_dfs = {}

for df_name, df in dfs.items():
  if 'response_group' in df.columns:
    grouped_dfs[df_name] = {}
    for group, group_df in df.groupby('response_group'):
      grouped_dfs[df_name][group] = group_df.copy() # Use copy() to avoid SettingWithCopyWarning
  else:
    print(f"DataFrame '{df_name}' does not have a 'response_group' column.")



In [ ]:
grouped_dfs['Response_survey_rows (2)']['allocation']

id                     created_at                               user_id  \
4    161  2025-05-30 10:57:33.488592+00  b308f105-ab1d-4bc0-9222-6ca1467f2da7   
5    162  2025-05-30 10:57:33.488592+00  b308f105-ab1d-4bc0-9222-6ca1467f2da7   
6    163  2025-05-30 10:57:33.488592+00  b308f105-ab1d-4bc0-9222-6ca1467f2da7   
7    164  2025-05-30 10:57:33.488592+00  b308f105-ab1d-4bc0-9222-6ca1467f2da7   
8    165  2025-05-30 11:11:22.184429+00  f568eacb-013d-4837-8da2-22a28f8538e6   
..   ...                            ...                                   ...   
135  292   2025-06-02 17:10:11.48677+00  98f0865c-4c7f-49b8-a376-49ba77f987a9   
148  305  2025-06-02 17:50:16.471356+00  a3300a28-6467-4a55-bfb3-0102e09ed761   
149  306  2025-06-02 17:50:16.471356+00  a3300a28-6467-4a55-bfb3-0102e09ed761   
150  307  2025-06-02 17:50:16.471356+00  a3300a28-6467-4a55-bfb3-0102e09ed761   
151  308  2025-06-02 17:50:16.471356+00  a3300a28-6467-4a55-bfb3-0102e09ed761   

                      response_project_id response_group  response_value  \
4    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               5   
5    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               6   
6    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               7   
7    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               5   
8    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               7   
..                                    ...            ...             ...   
135  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               5   
148  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               7   
149  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               5   
150  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               2   
151  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e     allocation               4   

     response_survey  
4                  5  
5                  6  
6                  7  
7                  8  
8                  5  
..               ...  
135                8  
148                5  
149                6  
150                7  
151                8  

[76 rows x 7 columns]

In [ ]:

allocated_survey=grouped_dfs['Response_survey_rows (2)']['allocation'].copy()
ranked_survey=grouped_dfs['Response_survey_rows (2)']['ranking'].copy()


In [ ]:
# prompt: box plot of allocated_survey response_value vs that of ranked survey when response_survey is 5 use plotly

import pandas as pd
!pip install plotly

import plotly.express as px

# Filter the allocated_survey and ranked_survey dataframes for response_survey == 5
allocated_survey_filtered = allocated_survey[allocated_survey['response_survey'] == 5]
ranked_survey_filtered = ranked_survey[ranked_survey['response_survey'] == 5]

# Combine the relevant data for plotting
plot_data = pd.DataFrame({
    'Allocation method': allocated_survey_filtered['response_value'],
    'Rankion method': ranked_survey_filtered['response_value']
})

# Melt the DataFrame for easier plotting with Plotly
melted_plot_data = plot_data.melt(value_vars=['Allocation method', 'Rankion method'],
                                  var_name='Survey Type',
                                  value_name='Fairness Score ')

# Create the box plot using Plotly Express
fig = px.box(melted_plot_data, y='Fairness Score ', color='Survey Type',
             title='Box Plot Showing Fairness Score of Allocation and Ranking Methods')

fig.show()


In [ ]:

import pandas as pd
!pip install plotly

import plotly.express as px

# Filter the allocated_survey and ranked_survey dataframes for response_survey == 6
allocated_survey_filtered = allocated_survey[allocated_survey['response_survey'] == 6]
ranked_survey_filtered = ranked_survey[ranked_survey['response_survey'] == 6]

# Combine the relevant data for plotting
plot_data = pd.DataFrame({
    'Allocated method': allocated_survey_filtered['response_value'],
    'Ranking method': ranked_survey_filtered['response_value']
})

# Melt the DataFrame for easier plotting with Plotly
melted_plot_data = plot_data.melt(value_vars=['Allocated method', 'Ranking method'],
                                  var_name='Survey Type',
                                  value_name='Expressiveness Score')

# Create the box plot using Plotly Express
fig = px.box(melted_plot_data, y='Expressiveness Score', color='Survey Type',
             title='Box Plot Showing the Expressiveness Score of Allocation and Ranking Methods')

fig.show()


In [ ]:

import pandas as pd
!pip install plotly

import plotly.express as px

# Filter the allocated_survey and ranked_survey dataframes for response_survey == 7
allocated_survey_filtered = allocated_survey[allocated_survey['response_survey'] == 7]
ranked_survey_filtered = ranked_survey[ranked_survey['response_survey'] == 7]

# Combine the relevant data for plotting
plot_data = pd.DataFrame({
    'Allocation method': allocated_survey_filtered['response_value'],
    'Ranking method': ranked_survey_filtered['response_value']
})

# Melt the DataFrame for easier plotting with Plotly
melted_plot_data = plot_data.melt(value_vars=['Allocation method', 'Ranking method'],
                                  var_name='Survey Type',
                                  value_name='Ease of Use Score')

# Create the box plot using Plotly Express
fig = px.box(melted_plot_data, y='Ease of Use Score', color='Survey Type',
             title='Box Plot Showing Ease of Use Score of Allocation and Ranking Methods')

fig.show()


In [ ]:

import pandas as pd
!pip install plotly

import plotly.express as px

# Filter the allocated_survey and ranked_survey dataframes for response_survey == 8
allocated_survey_filtered = allocated_survey[allocated_survey['response_survey'] == 8]
ranked_survey_filtered = ranked_survey[ranked_survey['response_survey'] == 8]

# Combine the relevant data for plotting
plot_data = pd.DataFrame({
    'Allocation method': allocated_survey_filtered['response_value'],
    'Ranking method': ranked_survey_filtered['response_value']
})

# Melt the DataFrame for easier plotting with Plotly
melted_plot_data = plot_data.melt(value_vars=['Allocation method', 'Ranking method'],
                                  var_name='Survey Type',
                                  value_name='Willingness to Reuse Score')

# Create the box plot using Plotly Express
fig = px.box(melted_plot_data, y='Willingness to Reuse Score', color='Survey Type',
             title='Box Plot Showing Willingness to Reuse Score of Allocation and Ranking Methods')

fig.show()


In [ ]:
# prompt: for 5,6,7,8 above tabulate the values of the mean and median for allcation and ranking

import pandas as pd
# Calculate mean and median for allocation and ranking for response_survey 5, 6, 7, 8
summary_data = []
for survey_value in [5, 6, 7, 8]:
  allocated_filtered = allocated_survey[allocated_survey['response_survey'] == survey_value]['response_value']
  ranked_filtered = ranked_survey[ranked_survey['response_survey'] == survey_value]['response_value']

  summary_data.append({
      'Response Survey Value': survey_value,
      'Allocation Mean': allocated_filtered.mean(),
      'Allocation Median': allocated_filtered.median(),
      'Ranking Mean': ranked_filtered.mean(),
      'Ranking Median': ranked_filtered.median()
  })

summary_df = pd.DataFrame(summary_data)

# Display the table
print(summary_df.to_markdown(index=False))

|   Response Survey Value |   Allocation Mean |   Allocation Median |   Ranking Mean |   Ranking Median |
|------------------------:|------------------:|--------------------:|---------------:|-----------------:|
|                       5 |           5.26316 |                   5 |        5.78947 |                6 |
|                       6 |           5.57895 |                   6 |        5.68421 |                6 |
|                       7 |           5.63158 |                   6 |        5.94737 |                6 |
|                       8 |           5.10526 |                   5 |        5.63158 |                5 |


In [ ]:
# prompt: create a pivot for allocated survey on user id with response survey

# Create a pivot table
# The index will be the 'user_id', the columns will be 'response_survey',
# and the values will be 'response_value'.
# We will use the first occurrence of 'response_value' for each user_id/response_survey combination.
pivot_table_allocation = allocated_survey.pivot_table(index='user_id',
                                           columns='response_survey',
                                           values='response_value')

pivot_table_ranking = ranked_survey.pivot_table(index='user_id',
                                           columns='response_survey',
                                           values='response_value')

print(pivot_table_ranking.head())
print(pivot_table_allocation.head())

response_survey                         5    6    7    8
user_id                                                 
004dd665-af00-4df8-8589-fc337e926e77  4.0  4.0  4.0  4.0
11c2caf2-a684-4591-96a8-7a54b353e022  4.0  5.0  5.0  5.0
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  6.0  6.0  5.0  5.0
2c96b7c9-5201-400e-9004-dc6e79730c79  5.0  7.0  5.0  5.0
381fb141-beed-4203-8882-053353b9266a  4.0  6.0  7.0  7.0
response_survey                         5    6    7    8
user_id                                                 
06358995-e955-4504-a830-26a31fecd93a  4.0  7.0  7.0  7.0
32f4d990-5657-4293-ad0d-1e2ee8614716  4.0  3.0  3.0  3.0
451c864a-8712-4977-bef8-2fb3c13a0af9  6.0  7.0  7.0  7.0
4a11a408-500f-464c-a7bd-f76631b1eaac  5.0  5.0  4.0  4.0
68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b  5.0  6.0  6.0  5.0


In [ ]:
# prompt: for pivot_table_ranking and pivot_table_allocation add a new column for score which is average across rows

pivot_table_ranking['score'] = pivot_table_ranking.mean(axis=1)
pivot_table_allocation['score'] = pivot_table_allocation.mean(axis=1)

print("\nPivot Table Ranking with Score:")
print(pivot_table_ranking.head())

print("\nPivot Table Allocation with Score:")
print(pivot_table_allocation.head())


Pivot Table Ranking with Score:
response_survey                         5    6    7    8  score
user_id                                                        
004dd665-af00-4df8-8589-fc337e926e77  4.0  4.0  4.0  4.0   4.00
11c2caf2-a684-4591-96a8-7a54b353e022  4.0  5.0  5.0  5.0   4.75
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  6.0  6.0  5.0  5.0   5.50
2c96b7c9-5201-400e-9004-dc6e79730c79  5.0  7.0  5.0  5.0   5.50
381fb141-beed-4203-8882-053353b9266a  4.0  6.0  7.0  7.0   6.00

Pivot Table Allocation with Score:
response_survey                         5    6    7    8  score
user_id                                                        
06358995-e955-4504-a830-26a31fecd93a  4.0  7.0  7.0  7.0   6.25
32f4d990-5657-4293-ad0d-1e2ee8614716  4.0  3.0  3.0  3.0   3.25
451c864a-8712-4977-bef8-2fb3c13a0af9  6.0  7.0  7.0  7.0   6.75
4a11a408-500f-464c-a7bd-f76631b1eaac  5.0  5.0  4.0  4.0   4.50
68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b  5.0  6.0  6.0  5.0   5.50


In [ ]:
# prompt: run a normality test on the score colu,n of each

from scipy import stats

print("\nNormality Test for 'score' in Pivot Table Ranking:")
# Drop any potential NaN values in the 'score' column before testing
ranking_score = pivot_table_ranking['score'].dropna()
shapiro_test_ranking = stats.shapiro(ranking_score)
print(f"Shapiro-Wilk Test (Ranking): Statistic={shapiro_test_ranking.statistic:.4f}, p-value={shapiro_test_ranking.pvalue:.4f}")

print("\nNormality Test for 'score' in Pivot Table Allocation:")
# Drop any potential NaN values in the 'score' column before testing
allocation_score = pivot_table_allocation['score'].dropna()
shapiro_test_allocation = stats.shapiro(allocation_score)
print(f"Shapiro-Wilk Test (Allocation): Statistic={shapiro_test_allocation.statistic:.4f}, p-value={shapiro_test_allocation.pvalue:.4f}")


Normality Test for 'score' in Pivot Table Ranking:
Shapiro-Wilk Test (Ranking): Statistic=0.9511, p-value=0.5785

Normality Test for 'score' in Pivot Table Allocation:
Shapiro-Wilk Test (Allocation): Statistic=0.9678, p-value=0.8237


In [ ]:
# prompt: run a two-tailed Welch’s independent t-test to compare both and include the sd  and CI

# Perform Welch's t-test (independent samples)
# This test does not assume equal variances, which is suitable if the Shapiro-Wilk test suggests non-normality or if variance is unequal.
ttest_result = stats.ttest_ind(ranking_score, allocation_score, equal_var=False, nan_policy='omit')

print("\nWelch's Independent t-test Results:")
print(f"Test Statistic (t): {ttest_result.statistic:.4f}")
print(f"P-value: {ttest_result.pvalue:.4f}")
print(f"Degrees of Freedom: {ttest_result.df:.2f}")

# Calculate descriptive statistics including standard deviation
ranking_mean = ranking_score.mean()
ranking_std = ranking_score.std()
allocation_mean = allocation_score.mean()
allocation_std = allocation_score.std()

print(f"\nDescriptive Statistics:")
print(f"Ranking Method: Mean={ranking_mean:.4f}, Std Dev={ranking_std:.4f}")
print(f"Allocation Method: Mean={allocation_mean:.4f}, Std Dev={allocation_std:.4f}")

# Calculate Confidence Interval for the difference in means
# We will use the formula for the confidence interval of the difference between two independent means (Welch-Satterthwaite)
# from statsmodels, which is more robust for unequal variances.
!pip install statsmodels
import statsmodels.stats.api as sms

# Calculate the confidence interval for the difference in means
# The confidence interval is for (mean_ranking - mean_allocation)
cm = sms.CompareMeans(sms.DescrStatsW(ranking_score), sms.DescrStatsW(allocation_score))
ci_diff = cm.tconfint_diff(usevar='unequal') # Welch's t-test uses unequal variances

print(f"\n95% Confidence Interval for the Difference in Means (Ranking - Allocation):")
print(f"Lower Bound: {ci_diff[0]:.4f}")
print(f"Upper Bound: {ci_diff[1]:.4f}")


Welch's Independent t-test Results:
Test Statistic (t): 0.9238
P-value: 0.3638
Degrees of Freedom: 26.95

Descriptive Statistics:
Ranking Method: Mean=5.7679, Std Dev=0.8741
Allocation Method: Mean=5.4500, Std Dev=0.9783

95% Confidence Interval for the Difference in Means (Ranking - Allocation):
Lower Bound: -0.3882
Upper Bound: 1.0239


In [ ]:
# prompt: do descriptive stats for both

# Display descriptive statistics for both score columns
print("\nDescriptive Statistics for 'score' in Pivot Table Ranking:")
print(pivot_table_ranking['score'].describe())

print("\nDescriptive Statistics for 'score' in Pivot Table Allocation:")
print(pivot_table_allocation['score'].describe())


Descriptive Statistics for 'score' in Pivot Table Ranking:
count    14.000000
mean      5.767857
std       0.874117
min       4.000000
25%       5.500000
50%       5.750000
75%       6.437500
max       7.000000
Name: score, dtype: float64

Descriptive Statistics for 'score' in Pivot Table Allocation:
count    15.000000
mean      5.450000
std       0.978337
min       3.250000
25%       4.875000
50%       5.500000
75%       6.000000
max       7.000000
Name: score, dtype: float64


In [ ]:
grouped_dfs['Response_rows (2)']['allocation'].head()

id                     created_at                               user_id  \
0  336  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
1  341  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
2  343  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
3  344  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
4  342  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   

                    response_project_id  response_policy_id response_group  \
0  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   1     allocation   
1  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   6     allocation   
2  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   8     allocation   
3  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   9     allocation   
4  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   7     allocation   

   response_value  response_bts_value  
0               5                  70  
1               2                  20  
2               3                  10  
3               0                   0  
4               0                   0

In [ ]:
dfs['Response_rows (2)']

id                     created_at                               user_id  \
0    336  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
1    341  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
2    343  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
3    344  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
4    342  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
..   ...                            ...                                   ...   
375  425  2025-05-31 13:11:35.744502+00  004dd665-af00-4df8-8589-fc337e926e77   
376  418  2025-05-31 13:11:35.744502+00  004dd665-af00-4df8-8589-fc337e926e77   
377  421  2025-05-31 13:11:35.744502+00  004dd665-af00-4df8-8589-fc337e926e77   
378  422  2025-05-31 13:11:35.744502+00  004dd665-af00-4df8-8589-fc337e926e77   
379  423  2025-05-31 13:11:35.744502+00  004dd665-af00-4df8-8589-fc337e926e77   

                      response_project_id  response_policy_id response_group  \
0    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   1     allocation   
1    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   6     allocation   
2    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   8     allocation   
3    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   9     allocation   
4    d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   7     allocation   
..                                    ...                 ...            ...   
375  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                  10        ranking   
376  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   3        ranking   
377  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   6        ranking   
378  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   7        ranking   
379  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   8        ranking   

     response_value  response_bts_value  
0                 5                  70  
1                 2                  20  
2                 3                  10  
3                 0                   0  
4                 0                   0  
..              ...                 ...  
375              10                  10  
376               3                   7  
377               6                   3  
378               7                   0  
379               8                   0  

[380 rows x 8 columns]

In [ ]:
# prompt: create bts df from dfs['Response_rows (2)'] keeping user_id, response_policy_id, response_bts_value

bts_df = dfs['Response_rows (2)'][['user_id', 'response_policy_id', 'response_bts_value']].copy()
bts_df.head()

user_id  response_policy_id  \
0  f568eacb-013d-4837-8da2-22a28f8538e6                   1   
1  f568eacb-013d-4837-8da2-22a28f8538e6                   6   
2  f568eacb-013d-4837-8da2-22a28f8538e6                   8   
3  f568eacb-013d-4837-8da2-22a28f8538e6                   9   
4  f568eacb-013d-4837-8da2-22a28f8538e6                   7   

   response_bts_value  
0                  70  
1                  20  
2                  10  
3                   0  
4                   0

In [ ]:
# prompt: create pivot from bts_df  on user_id

bts_pivot = bts_df.pivot_table(index='user_id',
                               columns='response_policy_id',
                               values='response_bts_value')

print(bts_pivot.head())

response_policy_id                      1     2     3     4     5     6    7   \
user_id                                                                         
004dd665-af00-4df8-8589-fc337e926e77  27.0  13.0   7.0  10.0  17.0   3.0  0.0   
06358995-e955-4504-a830-26a31fecd93a  30.0  20.0  20.0  10.0   5.0   5.0  5.0   
11c2caf2-a684-4591-96a8-7a54b353e022  20.0  10.0  40.0  10.0  20.0   0.0  0.0   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  20.0   0.0  30.0   0.0   0.0  20.0  0.0   
2c96b7c9-5201-400e-9004-dc6e79730c79   0.0  20.0   0.0   0.0  30.0   0.0  0.0   

response_policy_id                      8     9     10  
user_id                                                 
004dd665-af00-4df8-8589-fc337e926e77   0.0  13.0  10.0  
06358995-e955-4504-a830-26a31fecd93a   5.0   0.0   0.0  
11c2caf2-a684-4591-96a8-7a54b353e022   0.0   0.0   0.0  
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  30.0   0.0   0.0  
2c96b7c9-5201-400e-9004-dc6e79730c79   0.0   0.0   0.0  


In [ ]:
# prompt: create a new df bts_pivot contaning user_id ad sum of row

bts_pivot['Sum'] = bts_pivot.sum(axis=1)

print(bts_pivot.head())

response_policy_id                       1     2     3     4     5     6    7  \
user_id                                                                         
004dd665-af00-4df8-8589-fc337e926e77  27.0  13.0   7.0  10.0  17.0   3.0  0.0   
06358995-e955-4504-a830-26a31fecd93a  30.0  20.0  20.0  10.0   5.0   5.0  5.0   
11c2caf2-a684-4591-96a8-7a54b353e022  20.0  10.0  40.0  10.0  20.0   0.0  0.0   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  20.0   0.0  30.0   0.0   0.0  20.0  0.0   
2c96b7c9-5201-400e-9004-dc6e79730c79   0.0  20.0   0.0   0.0  30.0   0.0  0.0   

response_policy_id                       8     9    10    Sum  
user_id                                                        
004dd665-af00-4df8-8589-fc337e926e77   0.0  13.0  10.0  100.0  
06358995-e955-4504-a830-26a31fecd93a   5.0   0.0   0.0  100.0  
11c2caf2-a684-4591-96a8-7a54b353e022   0.0   0.0   0.0  100.0  
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  30.0   0.0   0.0  100.0  
2c96b7c9-5201-400e-9004-dc6e79730c79   0.0   0.0

In [ ]:
# prompt: bts_pivot['Sum'] for bts_pivot['Sum']ot equal to 100

rows_not_100 = bts_pivot[bts_pivot['Sum'] != 100]
print("\nRows where 'Sum' is not equal to 100:")
rows_not_100


Rows where 'Sum' is not equal to 100:


response_policy_id                       1     2     3     4     5     6  \
user_id                                                                    
2c96b7c9-5201-400e-9004-dc6e79730c79   0.0  20.0   0.0   0.0  30.0   0.0   
32f4d990-5657-4293-ad0d-1e2ee8614716  15.0  10.0  10.0  10.0  10.0  15.0   

response_policy_id                       7     8     9    10    Sum  
user_id                                                              
2c96b7c9-5201-400e-9004-dc6e79730c79   0.0   0.0   0.0   0.0   50.0  
32f4d990-5657-4293-ad0d-1e2ee8614716  10.0  10.0  10.0  25.0  125.0

In [ ]:
# prompt: subtract 2.5 for bts_pivot user_id 32f4d990-5657-4293-ad0d-1e2ee8614716 for 1,2,3,4,5,6,7,8,9,10 and check btspivot sum column

# Define the user_id and the values to subtract 2.5 from
user_id_to_update = '32f4d990-5657-4293-ad0d-1e2ee8614716'
columns_to_subtract = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Check if the user_id exists in the index
if user_id_to_update in bts_pivot.index:
  # Subtract 2.5 from the specified columns for the given user_id
  for col in columns_to_subtract:
    if col in bts_pivot.columns:
      bts_pivot.loc[user_id_to_update, col] -= 2.5

  # Recalculate the 'Sum' column for the updated row
  bts_pivot.loc[user_id_to_update, 'Sum'] = bts_pivot.loc[user_id_to_update, columns_to_subtract].sum()

  print(f"\nUpdated row for user_id: {user_id_to_update}")
  print(bts_pivot.loc[user_id_to_update])

  # Check the updated sum
  print(f"\nNew sum for user_id {user_id_to_update}: {bts_pivot.loc[user_id_to_update, 'Sum']}")

else:
  print(f"User ID '{user_id_to_update}' not found in the bts_pivot DataFrame.")

# You can optionally check all rows where the sum is not 100 again
rows_not_100_after_update = bts_pivot[bts_pivot['Sum'].round(2) != 100] # Use round for floating point comparisons
print("\nRows where 'Sum' is not equal to 100 after update:")
rows_not_100_after_update


Updated row for user_id: 32f4d990-5657-4293-ad0d-1e2ee8614716
response_policy_id
1       12.5
2        7.5
3        7.5
4        7.5
5        7.5
6       12.5
7        7.5
8        7.5
9        7.5
10      22.5
Sum    100.0
Name: 32f4d990-5657-4293-ad0d-1e2ee8614716, dtype: float64

New sum for user_id 32f4d990-5657-4293-ad0d-1e2ee8614716: 100.0

Rows where 'Sum' is not equal to 100 after update:


response_policy_id                      1     2    3    4     5    6    7  \
user_id                                                                     
2c96b7c9-5201-400e-9004-dc6e79730c79  0.0  20.0  0.0  0.0  30.0  0.0  0.0   

response_policy_id                      8    9   10   Sum  
user_id                                                    
2c96b7c9-5201-400e-9004-dc6e79730c79  0.0  0.0  0.0  50.0

In [ ]:

rows_not_100 = bts_pivot[bts_pivot['Sum'] != 100]
print("\nRows where 'Sum' is not equal to 100:")
rows_not_100


Rows where 'Sum' is not equal to 100:


response_policy_id                      1     2    3    4     5    6    7  \
user_id                                                                     
2c96b7c9-5201-400e-9004-dc6e79730c79  0.0  20.0  0.0  0.0  30.0  0.0  0.0   

response_policy_id                      8    9   10   Sum  
user_id                                                    
2c96b7c9-5201-400e-9004-dc6e79730c79  0.0  0.0  0.0  50.0

In [ ]:
# prompt: multiply 2   for bts_pivot user_id 2c96b7c9-5201-400e-9004-dc6e79730c79 for 1,2,3,4,5,6,7,8,9,10 and check btspivot sum column

# Define the user_id and the values to multiply by 2
user_id_to_update = '2c96b7c9-5201-400e-9004-dc6e79730c79'
columns_to_multiply = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Check if the user_id exists in the index
if user_id_to_update in bts_pivot.index:
  # Multiply the specified columns by 2 for the given user_id
  for col in columns_to_multiply:
    if col in bts_pivot.columns:
      bts_pivot.loc[user_id_to_update, col] *= 2

  # Recalculate the 'Sum' column for the updated row
  # Ensure to only sum the columns used for the original pivot calculation
  # If the original sum was calculated over all numeric columns, adjust as needed
  bts_pivot.loc[user_id_to_update, 'Sum'] = bts_pivot.loc[user_id_to_update, columns_to_multiply].sum()

  print(f"\nUpdated row for user_id: {user_id_to_update}")
  print(bts_pivot.loc[user_id_to_update])

  # Check the updated sum
  print(f"\nNew sum for user_id {user_id_to_update}: {bts_pivot.loc[user_id_to_update, 'Sum']}")

else:
  print(f"User ID '{user_id_to_update}' not found in the bts_pivot DataFrame.")

# Check all rows where the sum is not 100 again (using round for floating point comparisons)
rows_not_100_after_update = bts_pivot[bts_pivot['Sum'].round(2) != 100]
print("\nRows where 'Sum' is not equal to 100 after update:")
rows_not_100_after_update


Updated row for user_id: 2c96b7c9-5201-400e-9004-dc6e79730c79
response_policy_id
1        0.0
2       40.0
3        0.0
4        0.0
5       60.0
6        0.0
7        0.0
8        0.0
9        0.0
10       0.0
Sum    100.0
Name: 2c96b7c9-5201-400e-9004-dc6e79730c79, dtype: float64

New sum for user_id 2c96b7c9-5201-400e-9004-dc6e79730c79: 100.0

Rows where 'Sum' is not equal to 100 after update:


Empty DataFrame
Columns: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, Sum]
Index: []

In [ ]:
# prompt: sum of bts_pivot sum column

total_sum_of_bts_pivot_sum = bts_pivot['Sum'].sum()
print(f"\nTotal sum of the 'Sum' column in bts_pivot: {total_sum_of_bts_pivot_sum}")



Total sum of the 'Sum' column in bts_pivot: 2900.0


In [ ]:
# prompt:  average of bts_pivot['1]

average_of_col_1 = bts_pivot[1].mean()
print(f"\nAverage of column '1' in bts_pivot: {average_of_col_1}")

average_of_col_2 = bts_pivot[2].mean()
print(f"\nAverage of column '2' in bts_pivot: {average_of_col_2}")

average_of_col_3 = bts_pivot[3].mean()
print(f"\nAverage of column '3' in bts_pivot: {average_of_col_3}")

average_of_col_4 = bts_pivot[4].mean()
print(f"\nAverage of column '4' in bts_pivot: {average_of_col_4}")

average_of_col_5 = bts_pivot[5].mean()
print(f"\nAverage of column '5' in bts_pivot: {average_of_col_5}")

average_of_col_6 = bts_pivot[6].mean()
print(f"\nAverage of column '6' in bts_pivot: {average_of_col_6}")

average_of_col_7 = bts_pivot[7].mean()
print(f"\nAverage of column '7' in bts_pivot: {average_of_col_7}")

average_of_col_8 = bts_pivot[8].mean()
print(f"\nAverage of column '8' in bts_pivot: {average_of_col_8}")

average_of_col_9 = bts_pivot[9].mean()
print(f"\nAverage of column '9' in bts_pivot: {average_of_col_9}")

average_of_col_10 = bts_pivot[10].mean()
print(f"\nAverage of column '10' in bts_pivot: {average_of_col_10}")


Average of column '1' in bts_pivot: 28.844827586206897

Average of column '2' in bts_pivot: 9.879310344827585

Average of column '3' in bts_pivot: 14.637931034482758

Average of column '4' in bts_pivot: 6.810344827586207

Average of column '5' in bts_pivot: 13.258620689655173

Average of column '6' in bts_pivot: 11.293103448275861

Average of column '7' in bts_pivot: 2.8793103448275863

Average of column '8' in bts_pivot: 4.913793103448276

Average of column '9' in bts_pivot: 4.603448275862069

Average of column '10' in bts_pivot: 2.8793103448275863


In [ ]:
# prompt: plot the averages in a barchart and on the x axis do not skip any label

import pandas as pd
# Calculate the averages for columns 1 through 10
average_values = {}
for col_num in range(1, 11):
    average_values[col_num] = bts_pivot[col_num].mean()

# Create a DataFrame from the average values
average_df = pd.DataFrame(list(average_values.items()), columns=['Column', 'Average Value'])

# Create a bar chart using Plotly Express
fig = px.bar(average_df, x='Column', y='Average Value',
             title='Average Value for Each Column (1-10)')

# Ensure all x-axis labels are displayed
fig.update_layout(xaxis=dict(tickmode='linear'))

fig.show()

In [ ]:
# prompt: split allocation and ranking into 2 different df from dfs['Response_rows (2)']

# Split dfs['Response_rows (2)'] into allocation and ranking dataframes
allocation_df = dfs['Response_rows (2)'][dfs['Response_rows (2)']['response_group'] == 'allocation'].copy()
ranking_df = dfs['Response_rows (2)'][dfs['Response_rows (2)']['response_group'] == 'ranking'].copy()

print("Allocation DataFrame Head:")
print(allocation_df.head())

print("\nRanking DataFrame Head:")
print(ranking_df.head())

Allocation DataFrame Head:
    id                     created_at                               user_id  \
0  336  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
1  341  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
2  343  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
3  344  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   
4  342  2025-05-30 11:11:23.041067+00  f568eacb-013d-4837-8da2-22a28f8538e6   

                    response_project_id  response_policy_id response_group  \
0  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   1     allocation   
1  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   6     allocation   
2  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   8     allocation   
3  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   9     allocation   
4  d9d9496e-0bda-42f7-a1ab-7d8e766d3c6e                   7     allocation   

   response_value  response_b

In [ ]:
# prompt: pivot allocation_df and ranking_df on user_id  (columns=response_policy_id

# Pivot the allocation_df
allocation_pivot = allocation_df.pivot_table(index='user_id',
                                             columns='response_policy_id',
                                             values='response_value') # Assuming you want to pivot on 'response_value'

print("\nPivoted Allocation DataFrame Head:")
print(allocation_pivot.head())

# Pivot the ranking_df
ranking_pivot = ranking_df.pivot_table(index='user_id',
                                       columns='response_policy_id',
                                       values='response_value') # Assuming you want to pivot on 'response_value'

print("\nPivoted Ranking DataFrame Head:")
print(ranking_pivot.head())


Pivoted Allocation DataFrame Head:
response_policy_id                      1    2    3    4    5    6    7    8   \
user_id                                                                         
06358995-e955-4504-a830-26a31fecd93a   3.0  2.0  1.0  1.0  1.0  1.0  1.0  0.0   
32f4d990-5657-4293-ad0d-1e2ee8614716   2.0  1.0  1.0  1.0  0.0  2.0  1.0  1.0   
451c864a-8712-4977-bef8-2fb3c13a0af9   1.0  0.0  2.0  0.0  3.0  2.0  0.0  0.0   
4a11a408-500f-464c-a7bd-f76631b1eaac  10.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b   2.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   

response_policy_id                     9    10  
user_id                                         
06358995-e955-4504-a830-26a31fecd93a  0.0  0.0  
32f4d990-5657-4293-ad0d-1e2ee8614716  0.0  1.0  
451c864a-8712-4977-bef8-2fb3c13a0af9  2.0  0.0  
4a11a408-500f-464c-a7bd-f76631b1eaac  0.0  0.0  
68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b  1.0  0.0  

Pivoted Ranking DataFrame Head:
response_policy_id  

In [ ]:
# prompt: reverse code values in ranking_pivot

# Find the maximum ranking value in the ranking_pivot DataFrame
max_rank = ranking_pivot.max().max()

# Reverse the code values by subtracting them from max_rank + 1
ranking_pivot_reversed = max_rank + 1 - ranking_pivot

print("\nReversed Ranking DataFrame Head:")
print(ranking_pivot_reversed.head())


Reversed Ranking DataFrame Head:
response_policy_id                      1    2     3     4    5    6    7   \
user_id                                                                      
004dd665-af00-4df8-8589-fc337e926e77  10.0  9.0   8.0   7.0  6.0  5.0  4.0   
11c2caf2-a684-4591-96a8-7a54b353e022   8.0  6.0  10.0   7.0  9.0  5.0  4.0   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65   6.0  5.0  10.0   4.0  9.0  8.0  3.0   
2c96b7c9-5201-400e-9004-dc6e79730c79   4.0  8.0   6.0  10.0  9.0  5.0  7.0   
381fb141-beed-4203-8882-053353b9266a   7.0  3.0  10.0   4.0  8.0  9.0  5.0   

response_policy_id                     8    9    10  
user_id                                              
004dd665-af00-4df8-8589-fc337e926e77  3.0  2.0  1.0  
11c2caf2-a684-4591-96a8-7a54b353e022  3.0  2.0  1.0  
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  2.0  7.0  1.0  
2c96b7c9-5201-400e-9004-dc6e79730c79  3.0  2.0  1.0  
381fb141-beed-4203-8882-053353b9266a  1.0  6.0  2.0  


In [ ]:
# prompt: each of the values in ranking_pivot_reversed scaled to a rum sum of 10

# Scale each row of ranking_pivot_reversed so the sum is 10
# Avoid division by zero for rows where the sum is already 0
ranking_pivot_reversed_scaled = ranking_pivot_reversed.apply(lambda row: (row / row.sum()) * 10 if row.sum() != 0 else row, axis=1)

print("\nScaled and Reversed Ranking DataFrame Head (Sum of 10):")
print(ranking_pivot_reversed_scaled.head())

# Optional: Verify that the sum of each row is approximately 10
print("\nVerifying row sums:")
print(ranking_pivot_reversed_scaled.sum(axis=1).round(2).head())


Scaled and Reversed Ranking DataFrame Head (Sum of 10):
response_policy_id                          1         2         3         4   \
user_id                                                                        
004dd665-af00-4df8-8589-fc337e926e77  1.818182  1.636364  1.454545  1.272727   
11c2caf2-a684-4591-96a8-7a54b353e022  1.454545  1.090909  1.818182  1.272727   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  1.090909  0.909091  1.818182  0.727273   
2c96b7c9-5201-400e-9004-dc6e79730c79  0.727273  1.454545  1.090909  1.818182   
381fb141-beed-4203-8882-053353b9266a  1.272727  0.545455  1.818182  0.727273   

response_policy_id                          5         6         7         8   \
user_id                                                                        
004dd665-af00-4df8-8589-fc337e926e77  1.090909  0.909091  0.727273  0.545455   
11c2caf2-a684-4591-96a8-7a54b353e022  1.636364  0.909091  0.727273  0.545455   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65  1.636364  1.454545  0.54

In [ ]:
# prompt: concat ranking_pivot_reversed_scaled and allocation_pivot into a new df

import pandas as pd
concatenated_df = pd.concat([ranking_pivot_reversed_scaled, allocation_pivot], axis=0)

print("\nConcatenated DataFrame Head:")
print(concatenated_df)


Concatenated DataFrame Head:
response_policy_id                           1         2         3         4   \
user_id                                                                         
004dd665-af00-4df8-8589-fc337e926e77   1.818182  1.636364  1.454545  1.272727   
11c2caf2-a684-4591-96a8-7a54b353e022   1.454545  1.090909  1.818182  1.272727   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65   1.090909  0.909091  1.818182  0.727273   
2c96b7c9-5201-400e-9004-dc6e79730c79   0.727273  1.454545  1.090909  1.818182   
381fb141-beed-4203-8882-053353b9266a   1.272727  0.545455  1.818182  0.727273   
3958a762-5b20-4f71-ad7a-b56ce2be1284   1.636364  1.818182  1.454545  1.272727   
5cb55b24-f90f-435e-a9e5-ed27e8560690   1.090909  1.818182  1.636364  1.272727   
6b0da821-7955-4145-9f61-1b34c4504273   1.454545  1.636364  1.090909  1.818182   
6eba176c-2545-408e-934c-b9dc7c6e98a5   1.818182  1.636364  1.454545  1.272727   
87dfaf75-144f-40c8-ac6d-ad9c7ab9acc1   1.818182  0.545455  1.454545  0.181818  

In [ ]:
# prompt: multiply all values in concat by 10

concatenated_df_multiplied = concatenated_df * 10

print("\nConcatenated DataFrame with values multiplied by 10:")
concatenated_df_multiplied



Concatenated DataFrame with values multiplied by 10:


response_policy_id                            1          2          3   \
user_id                                                                  
004dd665-af00-4df8-8589-fc337e926e77   18.181818  16.363636  14.545455   
11c2caf2-a684-4591-96a8-7a54b353e022   14.545455  10.909091  18.181818   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65   10.909091   9.090909  18.181818   
2c96b7c9-5201-400e-9004-dc6e79730c79    7.272727  14.545455  10.909091   
381fb141-beed-4203-8882-053353b9266a   12.727273   5.454545  18.181818   
3958a762-5b20-4f71-ad7a-b56ce2be1284   16.363636  18.181818  14.545455   
5cb55b24-f90f-435e-a9e5-ed27e8560690   10.909091  18.181818  16.363636   
6b0da821-7955-4145-9f61-1b34c4504273   14.545455  16.363636  10.909091   
6eba176c-2545-408e-934c-b9dc7c6e98a5   18.181818  16.363636  14.545455   
87dfaf75-144f-40c8-ac6d-ad9c7ab9acc1   18.181818   5.454545  14.545455   
9d9c813f-909b-450a-ac06-3c921b59a642   12.727273   1.818182  18.181818   
a0c95a9f-4467-4375-8d73-6758677b2333   16.363636   7.272727   1.818182   
a742bd3e-ccc1-4d18-9115-4ec0c0fbb3b6   14.545455  12.727273  10.909091   
f299c546-a53c-4d62-9d6f-068409c316a8   12.727273  10.909091  18.181818   
06358995-e955-4504-a830-26a31fecd93a   30.000000  20.000000  10.000000   
32f4d990-5657-4293-ad0d-1e2ee8614716   20.000000  10.000000  10.000000   
451c864a-8712-4977-bef8-2fb3c13a0af9   10.000000   0.000000  20.000000   
4a11a408-500f-464c-a7bd-f76631b1eaac  100.000000   0.000000   0.000000   
68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b   20.000000  10.000000  10.000000   
6e3b176c-2125-4af7-b08e-b40a5ff64e7d   20.000000  20.000000  30.000000   
895efec8-7c06-409e-8a29-1c8dfa3ae7fd   20.000000   0.000000  30.000000   
8c67f4bf-d337-4669-af02-44347bcb6af4   20.000000  10.000000  10.000000   
98f0865c-4c7f-49b8-a376-49ba77f987a9   30.000000  30.000000  10.000000   
a3300a28-6467-4a55-bfb3-0102e09ed761   10.000000  10.000000  20.000000   
a813b3d1-8cac-48ef-956d-4ebe002296bc   10.000000   0.000000  20.000000   
b308f105-ab1d-4bc0-9222-6ca1467f2da7   20.000000  20.000000  30.000000   
cbdd63e3-91ce-4c71-9136-3814d4ea7993  100.000000   0.000000   0.000000   
f2e8ebd4-3f7f-43b6-97d6-b2fa7deb5edf   40.000000  10.000000  20.000000   
f568eacb-013d-4837-8da2-22a28f8538e6   50.000000   0.000000   0.000000   

response_policy_id                           4          5          6   \
user_id                                                                 
004dd665-af00-4df8-8589-fc337e926e77  12.727273  10.909091   9.090909   
11c2caf2-a684-4591-96a8-7a54b353e022  12.727273  16.363636   9.090909   
18da6fcf-6036-4e4f-a7b3-11507c6e4e65   7.272727  16.363636  14.545455   
2c96b7c9-5201-400e-9004-dc6e79730c79  18.181818  16.363636   9.090909   
381fb141-beed-4203-8882-053353b9266a   7.272727  14.545455  16.363636   
3958a762-5b20-4f71-ad7a-b56ce2be1284  12.727273  10.909091   9.090909   
5cb55b24-f90f-435e-a9e5-ed27e8560690  12.727273   9.090909  14.545455   
6b0da821-7955-4145-9f61-1b34c4504273  18.181818   9.090909  12.727273   
6eba176c-2545-408e-934c-b9dc7c6e98a5  12.727273  10.909091   9.090909   
87dfaf75-144f-40c8-ac6d-ad9c7ab9acc1   1.818182  10.909091   7.272727   
9d9c813f-909b-450a-ac06-3c921b59a642   9.090909  16.363636  14.545455   
a0c95a9f-4467-4375-8d73-6758677b2333  14.545455  18.181818  12.727273   
a742bd3e-ccc1-4d18-9115-4ec0c0fbb3b6   9.090909  16.363636  18.181818   
f299c546-a53c-4d62-9d6f-068409c316a8  14.545455  16.363636   9.090909   
06358995-e955-4504-a830-26a31fecd93a  10.000000  10.000000  10.000000   
32f4d990-5657-4293-ad0d-1e2ee8614716  10.000000   0.000000  20.000000   
451c864a-8712-4977-bef8-2fb3c13a0af9   0.000000  30.000000  20.000000   
4a11a408-500f-464c-a7bd-f76631b1eaac   0.000000   0.000000   0.000000   
68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b  10.000000  10.000000  10.000000   
6e3b176c-2125-4af7-b08e-b40a5ff64e7d  10.000000  20.000000   0.000000   
895efec8-7c06-409e-8a29-1c8dfa3ae7fd  20.000000   0.000000   0.000000   
8c67f4bf-d337-4669-af02-44

In [ ]:
# prompt: check row sum for each rows in concatenated_df_multiplied

# Check the row sum for each row in concatenated_df_multiplied
row_sums = concatenated_df_multiplied.sum(axis=1)

print("\nRow sums for concatenated_df_multiplied:")
row_sums


Row sums for concatenated_df_multiplied:


user_id
004dd665-af00-4df8-8589-fc337e926e77    100.0
11c2caf2-a684-4591-96a8-7a54b353e022    100.0
18da6fcf-6036-4e4f-a7b3-11507c6e4e65    100.0
2c96b7c9-5201-400e-9004-dc6e79730c79    100.0
381fb141-beed-4203-8882-053353b9266a    100.0
3958a762-5b20-4f71-ad7a-b56ce2be1284    100.0
5cb55b24-f90f-435e-a9e5-ed27e8560690    100.0
6b0da821-7955-4145-9f61-1b34c4504273    100.0
6eba176c-2545-408e-934c-b9dc7c6e98a5    100.0
87dfaf75-144f-40c8-ac6d-ad9c7ab9acc1    100.0
9d9c813f-909b-450a-ac06-3c921b59a642    100.0
a0c95a9f-4467-4375-8d73-6758677b2333    100.0
a742bd3e-ccc1-4d18-9115-4ec0c0fbb3b6    100.0
f299c546-a53c-4d62-9d6f-068409c316a8    100.0
06358995-e955-4504-a830-26a31fecd93a    100.0
32f4d990-5657-4293-ad0d-1e2ee8614716    100.0
451c864a-8712-4977-bef8-2fb3c13a0af9    100.0
4a11a408-500f-464c-a7bd-f76631b1eaac    100.0
68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b    100.0
6e3b176c-2125-4af7-b08e-b40a5ff64e7d    100.0
895efec8-7c06-409e-8a29-1c8dfa3ae7fd    100.0
8c67f4bf-d337-4669-af02-44347bcb6af4    100.0
98f0865c-4c7f-49b8-a376-49ba77f987a9    100.0
a3300a28-6467-4a55-bfb3-0102e09ed761    100.0
a813b3d1-8cac-48ef-956d-4ebe002296bc    100.0
b308f105-ab1d-4bc0-9222-6ca1467f2da7    100.0
cbdd63e3-91ce-4c71-9136-3814d4ea7993    100.0
f2e8ebd4-3f7f-43b6-97d6-b2fa7deb5edf    100.0
f568eacb-013d-4837-8da2-22a28f8538e6    100.0
dtype: float64

In [ ]:
# prompt: mean of sum concatenated_df_multiplied[x] for col1,2,3,4,5,6,7,8,9,10 with a loop

# Calculate the mean of the sum for each column (1 through 10)
for col_num in range(1, 11):
    col_name = col_num # Assuming column names are the numbers 1 through 10
    if col_name in concatenated_df_multiplied.columns:
        # Calculate the sum of the column first, then the mean of that sum (which is just the sum / number of rows)
        # Or, more directly, calculate the mean of the column
        mean_of_col_sum = concatenated_df_multiplied[col_name].sum() / len(concatenated_df_multiplied)
        print(f"Mean of sum for column {col_name}: {mean_of_col_sum}")
    else:
        print(f"Column {col_name} not found in concatenated_df_multiplied.")

# Alternatively, calculate the mean directly for each column:
print("\nCalculating mean for each column directly:")
for col_num in range(1, 11):
    col_name = col_num
    if col_name in concatenated_df_multiplied.columns:
        mean_value = concatenated_df_multiplied[col_name].mean()
        print(f"Mean for column {col_name}: {mean_value}")
    else:
        print(f"Column {col_name} not found in concatenated_df_multiplied.")

Mean of sum for column 1: 24.07523510971787
Mean of sum for column 2: 10.470219435736677
Mean of sum for column 3: 14.482758620689655
Mean of sum for column 4: 9.09090909090909
Mean of sum for column 5: 10.78369905956113
Mean of sum for column 6: 10.877742946708466
Mean of sum for column 7: 5.642633228840126
Mean of sum for column 8: 6.833855799373041
Mean of sum for column 9: 4.921630094043888
Mean of sum for column 10: 2.8213166144200623

Calculating mean for each column directly:
Mean for column 1: 24.07523510971787
Mean for column 2: 10.470219435736677
Mean for column 3: 14.482758620689655
Mean for column 4: 9.09090909090909
Mean for column 5: 10.78369905956113
Mean for column 6: 10.877742946708466
Mean for column 7: 5.642633228840126
Mean for column 8: 6.833855799373041
Mean for column 9: 4.921630094043888
Mean for column 10: 2.8213166144200623


from BTS
Average of column '1' in bts_pivot: 28.844827586206897

Average of column '2' in bts_pivot: 9.879310344827585

Average of column '3' in bts_pivot: 14.637931034482758

Average of column '4' in bts_pivot: 6.810344827586207

Average of column '5' in bts_pivot: 13.258620689655173

Average of column '6' in bts_pivot: 11.293103448275861

Average of column '7' in bts_pivot: 2.8793103448275863

Average of column '8' in bts_pivot: 4.913793103448276

Average of column '9' in bts_pivot: 4.603448275862069

Average of column '10' in bts_pivot: 2.8793103448275863

FROM NORMAL

---


Mean for column 1: 24.07523510971787

---


Mean for column 2: 10.470219435736677

---


Mean for column 3: 14.482758620689655

---


Mean for column 4: 9.09090909090909

---


Mean for column 5: 10.78369905956113

---


Mean for column 6: 10.877742946708466

---


Mean for column 7: 5.642633228840126

---


Mean for column 8: 6.833855799373041

---


Mean for column 9: 4.921630094043888

---


Mean for column 10: 2.8213166144200623

Policy 2 is both "above 10" and "surprisingly popular." This is a strong indicator: it's highly liked, AND its popularity was underestimated. This is a very strong candidate for a successful policy.

---


Policies 1, 3, 5, 6 are "above 10" but not "surprisingly popular" (in fact, their popularity was slightly overestimated). These are well-liked policies, but their high level of support might be more widely known or expected by the group.

---


Policies 4, 7, 8, 9 are "surprisingly popular" but not "above 10." These are policies whose support was underestimated, even though their absolute level of support isn't as high as the top tier. This suggests they might be niche favorites, or perhaps policies that face a perception problem (people think they're less popular than they are).

In [ ]:
[ranking_pivot_reversed_scaled, allocation_pivot]

[response_policy_id                          1         2         3         4   \
 user_id                                                                        
 004dd665-af00-4df8-8589-fc337e926e77  1.818182  1.636364  1.454545  1.272727   
 11c2caf2-a684-4591-96a8-7a54b353e022  1.454545  1.090909  1.818182  1.272727   
 18da6fcf-6036-4e4f-a7b3-11507c6e4e65  1.090909  0.909091  1.818182  0.727273   
 2c96b7c9-5201-400e-9004-dc6e79730c79  0.727273  1.454545  1.090909  1.818182   
 381fb141-beed-4203-8882-053353b9266a  1.272727  0.545455  1.818182  0.727273   
 3958a762-5b20-4f71-ad7a-b56ce2be1284  1.636364  1.818182  1.454545  1.272727   
 5cb55b24-f90f-435e-a9e5-ed27e8560690  1.090909  1.818182  1.636364  1.272727   
 6b0da821-7955-4145-9f61-1b34c4504273  1.454545  1.636364  1.090909  1.818182   
 6eba176c-2545-408e-934c-b9dc7c6e98a5  1.818182  1.636364  1.454545  1.272727   
 87dfaf75-144f-40c8-ac6d-ad9c7ab9acc1  1.818182  0.545455  1.454545  0.181818   
 9d9c813f-909b-450a-ac06-3c9

In [ ]:
for col_num in range(1, 11):
    col_name = col_num # Assuming column names are the numbers 1 through 10
    if col_name in ranking_pivot_reversed_scaled.columns:
        # Calculate the sum of the column first, then the mean of that sum (which is just the sum / number of rows)
        # Or, more directly, calculate the mean of the column
        mean_of_col_sum = ranking_pivot_reversed_scaled[col_name].sum() / len(ranking_pivot_reversed_scaled)
        print(f"Mean of sum for column {col_name}: {mean_of_col_sum}")
    else:
        print(f"Column {col_name} not found in concatenated_df_multiplied.")

Mean of sum for column 1: 1.4155844155844155
Mean of sum for column 2: 1.1688311688311688
Mean of sum for column 3: 1.4285714285714286
Mean of sum for column 4: 1.1688311688311686
Mean of sum for column 5: 1.3766233766233764
Mean of sum for column 6: 1.1818181818181817
Mean of sum for column 7: 0.7402597402597403
Mean of sum for column 8: 0.5584415584415584
Mean of sum for column 9: 0.5194805194805193
Mean of sum for column 10: 0.4415584415584415


In [ ]:
for col_num in range(1, 11):
    col_name = col_num # Assuming column names are the numbers 1 through 10
    if col_name in  allocation_pivot.columns:
        # Calculate the sum of the column first, then the mean of that sum (which is just the sum / number of rows)
        # Or, more directly, calculate the mean of the column
        mean_of_col_sum =  allocation_pivot[col_name].sum() / len( allocation_pivot)
        print(f"Mean of sum for column {col_name}: {mean_of_col_sum}")
    else:
        print(f"Column {col_name} not found in concatenated_df_multiplied.")

Mean of sum for column 1: 3.3333333333333335
Mean of sum for column 2: 0.9333333333333333
Mean of sum for column 3: 1.4666666666666666
Mean of sum for column 4: 0.6666666666666666
Mean of sum for column 5: 0.8
Mean of sum for column 6: 1.0
Mean of sum for column 7: 0.4
Mean of sum for column 8: 0.8
Mean of sum for column 9: 0.4666666666666667
Mean of sum for column 10: 0.13333333333333333


In [ ]:
# prompt: [ranking_pivot_reversed_scaled.loc[2], allocation_pivot.loc[2]] into 2 sets of new dfs

import pandas as pd
ranking_df2 = pd.DataFrame([ranking_pivot_reversed_scaled[2]])
allocation_df2 = pd.DataFrame([allocation_pivot[2]])

In [ ]:
# prompt: multiply the values in ranking_df2 and  allocation_df2 by 10

ranking_df2_multiplied = ranking_df2 * 10
allocation_df2_multiplied = allocation_df2 * 10

print("\nRanking DataFrame 2 with values multiplied by 10:")
print(ranking_df2_multiplied)

print("\nAllocation DataFrame 2 with values multiplied by 10:")
allocation_df2_multiplied


Ranking DataFrame 2 with values multiplied by 10:
user_id  004dd665-af00-4df8-8589-fc337e926e77  \
2                                   16.363636   

user_id  11c2caf2-a684-4591-96a8-7a54b353e022  \
2                                   10.909091   

user_id  18da6fcf-6036-4e4f-a7b3-11507c6e4e65  \
2                                    9.090909   

user_id  2c96b7c9-5201-400e-9004-dc6e79730c79  \
2                                   14.545455   

user_id  381fb141-beed-4203-8882-053353b9266a  \
2                                    5.454545   

user_id  3958a762-5b20-4f71-ad7a-b56ce2be1284  \
2                                   18.181818   

user_id  5cb55b24-f90f-435e-a9e5-ed27e8560690  \
2                                   18.181818   

user_id  6b0da821-7955-4145-9f61-1b34c4504273  \
2                                   16.363636   

user_id  6eba176c-2545-408e-934c-b9dc7c6e98a5  \
2                                   16.363636   

user_id  87dfaf75-144f-40c8-ac6d-ad9c7ab9acc1  \
2        

user_id  06358995-e955-4504-a830-26a31fecd93a  \
2                                        20.0   

user_id  32f4d990-5657-4293-ad0d-1e2ee8614716  \
2                                        10.0   

user_id  451c864a-8712-4977-bef8-2fb3c13a0af9  \
2                                         0.0   

user_id  4a11a408-500f-464c-a7bd-f76631b1eaac  \
2                                         0.0   

user_id  68e3f6ae-9779-4ab9-ab08-0ef4cef2c87b  \
2                                        10.0   

user_id  6e3b176c-2125-4af7-b08e-b40a5ff64e7d  \
2                                        20.0   

user_id  895efec8-7c06-409e-8a29-1c8dfa3ae7fd  \
2                                         0.0   

user_id  8c67f4bf-d337-4669-af02-44347bcb6af4  \
2                                        10.0   

user_id  98f0865c-4c7f-49b8-a376-49ba77f987a9  \
2                                        30.0   

user_id  a3300a28-6467-4a55-bfb3-0102e09ed761  \
2                                        10.0   

user_id  a813b3d1-8cac-48ef-956d-4ebe002296bc  \
2                                         0.0   

user_id  b308f105-ab1d-4bc0-9222-6ca1467f2da7  \
2                                        20.0   

user_id  cbdd63e3-91ce-4c71-9136-3814d4ea7993  \
2                                         0.0   

user_id  f2e8ebd4-3f7f-43b6-97d6-b2fa7deb5edf  \
2                                        10.0   

user_id  f568eacb-013d-4837-8da2-22a28f8538e6  
2                                         0.0

In [ ]:
# prompt: abs(allocation_df2_multiplied values- 9.879310344827585) and abs(ranking_df2_multiplied values- 9.879310344827585) in new df for all values

import pandas as pd
# Calculate the absolute difference from 9.879310344827585 for each value
ranking_diff_df = ranking_df2_multiplied.applymap(lambda x: (x - 9.879310344827585))
allocation_diff_df = allocation_df2_multiplied.applymap(lambda x: (x - 9.879310344827585))



/tmp/ipython-input-78-3075314810.py:5: FutureWarning:

DataFrame.applymap has been deprecated. Use DataFrame.map instead.

/tmp/ipython-input-78-3075314810.py:6: FutureWarning:

DataFrame.applymap has been deprecated. Use DataFrame.map instead.



In [ ]:
# prompt: do shapriro for each of ranking_diff_df and allocation_diff_df

print("\nShapiro-Wilk Test for error on ranking method:")
# Shapiro-Wilk test expects a 1D array, so we need to flatten the DataFrame
shapiro_test_ranking_diff = stats.shapiro(ranking_diff_df.values.flatten())
print(f"Shapiro-Wilk Test (error on ranking method): Statistic={shapiro_test_ranking_diff.statistic:.4f}, p-value={shapiro_test_ranking_diff.pvalue:.4f}")

print("\nShapiro-Wilk Test for error on allocation method")
# Shapiro-Wilk test expects a 1D array, so we need to flatten the DataFrame
shapiro_test_allocation_diff = stats.shapiro(allocation_diff_df.values.flatten())
print(f"Shapiro-Wilk Test ( for error on allocation method): Statistic={shapiro_test_allocation_diff.statistic:.4f}, p-value={shapiro_test_allocation_diff.pvalue:.4f}")


Shapiro-Wilk Test for error on ranking method:
Shapiro-Wilk Test (error on ranking method): Statistic=0.9305, p-value=0.3101

Shapiro-Wilk Test for error on allocation method
Shapiro-Wilk Test ( for error on allocation method): Statistic=0.8441, p-value=0.0143


In [ ]:
# prompt: run Mann-Whitney U test for them

# Perform Mann-Whitney U test
# This test is a non-parametric alternative to the independent samples t-test
# It is suitable when the assumption of normality is violated or when dealing with ordinal data.
# The inputs to the test are the two independent samples.

# We use the absolute difference values calculated previously
mannwhitneyu_result = stats.mannwhitneyu(ranking_diff_df.values.flatten(),
                                         allocation_diff_df.values.flatten(),
                                         alternative='two-sided') # 'two-sided' tests if the distributions are different

print("\nMann-Whitney U Test Results:")
print(f"Statistic (U): {mannwhitneyu_result.statistic:.4f}")
print(f"P-value: {mannwhitneyu_result.pvalue:.4f}")

# Interpret the p-value:
alpha = 0.05
if mannwhitneyu_result.pvalue < alpha:
    print(f"At the {alpha} significance level, we reject the null hypothesis.")
    print("There is a statistically significant difference in the distributions of errors between the ranking and allocation methods.")
else:
    print(f"At the {alpha} significance level, we fail to reject the null hypothesis.")
    print("There is not enough statistical evidence to conclude a difference in the distributions of errors between the ranking and allocation methods.")


Mann-Whitney U Test Results:
Statistic (U): 129.0000
P-value: 0.3011
At the 0.05 significance level, we fail to reject the null hypothesis.
There is not enough statistical evidence to conclude a difference in the distributions of errors between the ranking and allocation methods.


In [ ]:
# prompt: average value for ranking_diff_df, average value for allocation_diff_df

# Calculate the average (mean) of the values in ranking_diff_df
average_ranking_diff = ranking_diff_df.values.mean()
print(f"\nAverage error value for ranking_diff_df: {average_ranking_diff}")

# Calculate the average (mean) of the values in allocation_diff_df
average_allocation_diff = allocation_diff_df.values.mean()
print(f"Average error value for allocation_diff_df: {average_allocation_diff}")


Average error value for ranking_diff_df: 1.8090013434841026
Average error value for allocation_diff_df: -0.5459770114942522
